### PR TITLE
Typescript Migration: admin blink component

### PR DESCRIPTION
**Pull request by markxiong0122**
migrated the following component:

1. chromedash-admin-component-listing
2. chromedash-admin-component-listing_test
3. chromedash-admin-blink-page
4. chromedash-admin-blink-page_test

### CODE REVIEW

File: client-src/elements/chromedash-admin-blink-component-listing.ts [hunk 1]
```diff
@@ -1,8 +1,19 @@
...
+import {html, css, LitElement, TemplateResult} from 'lit';
 import {SHARED_STYLES} from '../css/shared-css.js';
 import {VARS} from '../css/_vars-css.js';
 import {LAYOUT_CSS} from '../css/_layout-css.js';
+import {customElement, property, query} from 'lit/decorators.js';
+import '../js-src/openapi-client';
+import {DefaultApiInterface} from 'chromestatus-openapi';
+import {get} from 'http';
 
+interface User {
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-component-listing.ts:10]: From the call site (https://github.com/GoogleChrome/chromium-dashboard/blob/e497b9f5b42603d4bc5e9a5a2a6fcbd62453cdb8/client-src/elements/chromedash-admin-blink-page.js#L79-L80), it looks like this is `import type {ComponentsUser} from 'chromestatus-openapi;`

File: client-src/elements/chromedash-admin-blink-component-listing_test.ts [hunk 2]
```diff
@@ -4,8 +4,10 @@ import {assert, expect, fixture, oneEvent} from '@open-wc/testing';
 import sinon from 'sinon';
 import {DefaultApi} from 'chromestatus-openapi';
 
-/** @type {Map<number, import('chromestatus-openapi').ComponentsUser>} */
-const testUsersMap = new Map([
+const testUsersMap = new Map<
+  number,
+  import('chromestatus-openapi').ComponentsUser
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-component-listing_test.ts:9]: Move this import up to the file level as `import type {ComponentsUser} from 'chromestatus-openapi';`

File: client-src/elements/chromedash-admin-blink-page.ts [hunk 3]
```diff
@@ -62,29 +65,21 @@ export class ChromedashAdminBlinkPage extends LitElement {
...
+  user = null;
+  @property({type: Boolean})
+  _editMode = false;
+  @property({
+    type: Array<import('chromestatus-openapi').OwnersAndSubscribersOfComponent>,
+  })
   components;
-
-  /** @type {Map<int, import('chromestatus-openapi').ComponentsUser}> */
+  @property({type: Map<number, import('chromestatus-openapi').ComponentsUser>})
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-page.ts:79]: Move these imports up to the file level.

File: client-src/elements/chromedash-admin-blink-page.ts [hunk 4]
```diff
@@ -191,8 +186,8 @@ export class ChromedashAdminBlinkPage extends LitElement {
     if (location.hash) {
       const hash = decodeURIComponent(location.hash);
       if (hash) {
-        const el = this.shadowRoot.querySelector(hash);
-        el.scrollIntoView(true, {behavior: 'smooth'});
+        const el = this.shadowRoot!.querySelector(hash);
+        el!.scrollIntoView({behavior: 'smooth'});
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-page.ts:190]: This could be a behavior change, but it's probably the intended behavior.

File: client-src/elements/chromedash-admin-blink-component-listing.ts [hunk 5]
```diff
@@ -81,49 +92,51 @@ export class ChromedashAdminBlinkComponentListing extends LitElement {
...
+  @property({type: Boolean, reflect: true})
+  editing = false;
+  @property({type: Object})
+  component = {};
+  @property({type: Number})
+  index = 0;
+  @property({type: Object})
+  usersMap: Map<number, User> = new Map();
+  @property({type: Number})
+  componentId = 0;
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-component-listing.ts:105]: Why did you rename `id` to `componentId`? It's not necessarily wrong, but surprising as part of the Typescript refactoring.

File: client-src/elements/chromedash-admin-blink-page.ts [hunk 6]
```diff
@@ -191,8 +186,8 @@ export class ChromedashAdminBlinkPage extends LitElement {
...
       }
     }
   }
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-page.ts:193]: You missed a use of `<chromedash-admin-blink-component-listing .id=...` in the section below this that won't match up with your renamed `.componentId` property.

File: client-src/elements/chromedash-admin-blink-page.ts (hunk 4 above)
Comment (markxiong0122) [reply client-src/elements/chromedash-admin-blink-page.ts:190]: yeah it only accepts 1 argument, and should be the configs

File: client-src/elements/chromedash-admin-blink-component-listing.ts (hunk 5 above)
Comment (markxiong0122) [reply client-src/elements/chromedash-admin-blink-component-listing.ts:105]: it was because there's user.id down below, and TypeScript does not allow this overlap

File: client-src/elements/chromedash-admin-blink-page.ts (hunk 6 above)
Comment (markxiong0122) [reply client-src/elements/chromedash-admin-blink-page.ts:193]: good catch. thanks

File: client-src/elements/chromedash-admin-blink-component-listing.ts (hunk 5 above)
Comment (markxiong0122) [reply client-src/elements/chromedash-admin-blink-component-listing.ts:99]: oh sorry i messed this up with another error, I changed this because `Property 'id' in type 'ChromedashAdminBlinkComponentListing' is not assignable to the same property in base type 'LitElement'.` as it extends from litelement

Comment (jyasskin) [reply client-src/elements/chromedash-admin-blink-component-listing.ts:99]: Aha! Because LitElement is an HTMLElement, and `id` there is https://developer.mozilla.org/en-US/docs/Web/API/Element/id. Yep, good change.

File: client-src/elements/chromedash-admin-blink-page.ts [hunk 7]
```diff
@@ -203,7 +202,7 @@ export class ChromedashAdminBlinkPage extends LitElement {
           (component, index) => html`
             <li class="layout horizontal" id="${component.name}">
               <chromedash-admin-blink-component-listing
-                .id=${component.id}
+                .id=${component.componentId}
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-page.ts:205]: When you change the name of an attribute or property in an element, you have to update the attributes and property assignments in uses of that element. You also need to make sure that when you access a field on an object, that field actually exists: `component` is an `OwnersAndSubscribersOfComponent`, which has an `id` field and not a `componentId` field. So this needs to change to

```suggestion
                .componentId=${component.id}
```

You also need to find a way to test this and any other uses of the property, to be sure you've correctly understood what needs to match. @jrobbins, what's a good way for Mark to test this?

File: client-src/elements/chromedash-admin-blink-page.ts [hunk 8]
```diff
@@ -62,29 +69,21 @@ export class ChromedashAdminBlinkPage extends LitElement {
...
+  @property({type: Boolean})
+  loading = true;
+  @property({type: Object})
+  user = null;
+  @property({type: Boolean})
+  _editMode = false;
+  @property({
+    type: Array<OwnersAndSubscribersOfComponent>,
+  })
   components;
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-page.ts:82]: This syntax doesn't work to tell Typescript what type a field has: https://lit.dev/docs/components/properties/#property-options:~:text=reactive%20state.-,type,-When%20converting%20a

The VSCode `runem.lit-plugin` extension notices this and underlines the incorrect type in yellow. Do you have that installed?

You'll want
```suggestion
  @state()
  components: Array<OwnersAndSubscribersOfComponent> | undefined;
```

(I changed `@property({attribute:false})` to `@state()` because these fields are assigned in `fetchData()` rather than in a surrounding template.)

File: client-src/elements/chromedash-admin-blink-page.ts [hunk 9]
```diff
@@ -62,29 +69,21 @@ export class ChromedashAdminBlinkPage extends LitElement {
...
-
-  /** @type {Map<int, import('chromestatus-openapi').ComponentsUser}> */
+  @property({type: Map<number, ComponentsUser>})
   usersMap;
```
Comment (jyasskin) [inline client-src/elements/chromedash-admin-blink-page.ts:84]: Similar to `components`, but without the `|undefined`
```suggestion
  @state()
  usersMap!: Map<number, ComponentsUser>;
```

**Review by jyasskin — CHANGES_REQUESTED** [review]

File: client-src/elements/chromedash-admin-blink-page.ts (hunk 8 above)
Comment (markxiong0122) [reply client-src/elements/chromedash-admin-blink-page.ts]: this results in a lot of `object might be undefined` errors. I manually added ! after each but not sure if there's a better way to do so

Comment (jyasskin) [reply client-src/elements/chromedash-admin-blink-page.ts]: I think the `!`s are correct for this, at least for this change. We might want to add some more explicit conditions to handle when `listComponentUsers` returns an undefined `components` field, or it might be that https://github.com/GoogleChrome/chromium-dashboard/blob/1b192c0ecfbb4f0e1eda57c4387bd9b7df5c8bbb/openapi/api.yaml#L192-L201 should change to list its properties as required (@jcscottiii?), but we can do either of those in a future PR.

Comment (jcscottiii) [reply client-src/elements/chromedash-admin-blink-page.ts]: We can definitely do that given that those two keys are always returned

https://github.com/GoogleChrome/chromium-dashboard/blob/1b192c0ecfbb4f0e1eda57c4387bd9b7df5c8bbb/api/components_users.py#L63-L64

Comment (jyasskin) [reply client-src/elements/chromedash-admin-blink-page.ts]: And the syntax for that is in https://swagger.io/docs/specification/data-models/dictionaries/#fixed-keys / https://json-schema.org/learn/getting-started-step-by-step#define-required-properties.

**Review by DanielRyanSmith — APPROVED** [review]

**Comment by markxiong0122** [comment]
migrated the following component:

chromedash-admin-component-listing
chromedash-admin-component-listing_test
chromedash-admin-blink-page
chromedash-admin-blink-page_test
